Tweaks for representation of the editor in the accessibility tree

Comandeer opened the pull request:
<!--
🚨 If you want to submit a PR for a security vulnerability, please contact us directly
at https://ckeditor.com/contact/ instead. 🚨
-->
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
Fixed Issues:

* [#4052](https://github.com/ckeditor/ckeditor4/issues/4052): Fixed: Editor labels are read incorrectly by screen readers due to invalid editor control type for [Iframe Editing Area](https://ckeditor.com/cke4/addon/wysiwygarea) editors.
* [#1904](https://github.com/ckeditor/ckeditor4/issues/1904): Fixed: Screen readers are not announcing read-only editor state due to missing proper aria-labels.
* [#2445](https://github.com/ckeditor/ckeditor4/issues/2445): Fixed: The same editor's label is read several times by screen readers when focusing an editor.

API Changes:

* [#2445](https://github.com/ckeditor/ckeditor4/issues/2445): Added [`config.applicationTitle`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-applicationTitle) alongside [`CKEDITOR.editor#applicationTitle`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#property-title) to allow customizing editor's application region label.
```

## What changes did you make?

I've added missing `[role=textbox]` and `[aria-multiline]` attributes for `wysiwygarea` editors. I've also added `[aria-readonly]` attribute to editors to indicate the status of the "readonliness" of the editor. However, for iframe-based editors it required adding also the `[tabindex=0]` attribute to be actually read by screen readers. This change was implemented for all browsers except Firefox which has a [bug connected with this attribute and `iframe`s](https://bugzilla.mozilla.org/show_bug.cgi?id=1483828). It also means that marking read-only `iframe`s does not work in Firefox.

I didn't touch #2445 because I have no idea how to differentiate labels for applications and editables. But the changes introduced for read-only editors indicate such a need (NVDA now reads the same label _thrice_ – probably due to it being the label of the application, the label of the editor, and the title of the editor's `iframe`). Probably it's something that we need to discuss further.

Proposed changes should also fix #2499 but it needs further verification.

## Which issues does your PR resolve?

Closes #4052.
Closes #1904.
Closes #2445.

jacekbogdanski reviewed (commented):
## NVDA

First thing is that I'm not actually able to reproduce https://github.com/ckeditor/ckeditor4/issues/4052
> Control type is not informational (I believe it is announced as 'group')

I've downloaded the latest version of NVDA, and it seems to properly read the editor labels (putting aside fact that the editor labels seem to be read a bit redundantly). So it's hard to tell me how these changes impacted reading labels as I don't see that many differences, except of course read-only support. @Comandeer did you manage to reproduce the original issue?

Currently, NVDA will read focused editor as:

Rich Text Editor, [editor name]
Rich Text Editor, [editor name] Frame
Rich Text Editor, [editor name] Document
Edit Multi Line [text]

In the case of a11yhelp plugin:

Rich Text Editor, [editor name]
Rich Text Editor, [editor name] Frame, press Alt + 0 for help
Rich Text Editor, [editor name] Document
Edit Multi Line [text]

I'm wondering if that much information is really needed. I'm not an accessibility expert, but wouldn't be more helpful to just read something like:

Rich Text Editor, [editor name]
Editing Area
[text]

In the case of a11yhelp plugin:
Rich Text Editor, [editor name], press Alt + 0 for help
Editing Area
[text]

It seems more helpful to skip that much information for me but maybe missing some important cases here. @Comandeer could you help me to understand why we are giving so much information here and duplicating the message?

I suppose that it may be related to switching between another focusable element (like toolbar) in the editor, but in that case, reading Editing Area would still give the user feedback on what's happening.

## VoiceOver
I didn't notice any difference except for different reading for read-only (which I suppose is good). It seems that VoiceOver doesn't care about reading Iframe titles at all.

## JAWS
Running JAWS resulted in freezing my VM, so I'm going to get back to it soon.

jacekbogdanski commented:
To sum up our discussion with @Comandeer F2F, we will try to come up with a simplified version of these labels:

Rich Text Editor, [editor name]
Editor [editor name] Frame, press Alt + 0 for help
Editing Area (let's check if this label can be skipped without making the editor less accessible)
[text]

It should reduce repetitive text while keeping a good level of accessibility for the different parts of the editor. Note that labels may be slightly different depending on results from another code iteration.

Comandeer commented:
Labels were far more difficult than I expected. I had to introduce `editor#applicationTitle` property alongside `CKEDITOR.config#applicationTitle` variable (as counterparts to `editor#title` and `CKEDITOR.config#title`). Additionally, I had to… add yet another label (`[aria-label]` attribute to the iframe's `body` for VoiceOver to correctly read the editor).

So there are two distinctive labels:

1.  application one ("Rich Text Editor, \<name>"),
2.  editor one ("Editor, \<name>").

The second one is used on editables and `iframe`s. Unfortunately, this label is repeated twice in NVDA and JAWS (once for the frame itself and once for the document inside it) but it's probably the best approach.

jacekbogdanski commented on tests/core/editable/manual/ariareadonlyscreenreader.md:
```diff
@@ -0,0 +1,16 @@
+@bender-tags: 4.18.1, bug, 1904
```
```suggestion
@bender-tags: 4.19.0, bug, 1904
```

jacekbogdanski commented on tests/plugins/wysiwygarea/manual/ariareadonlyscreenreader.md:
```diff
@@ -0,0 +1,16 @@
+@bender-tags: 4.18.1, bug, 1904
```
```suggestion
@bender-tags: 4.19.0, bug, 1904
```

jacekbogdanski commented on tests/plugins/wysiwygarea/manual/tabindex.md:
```diff
@@ -0,0 +1,11 @@
+@bender-tags: 4.18.1, bug, 1904
```
```suggestion
@bender-tags: 4.19.0, bug, 1904
```

jacekbogdanski reviewed (changes requested):
I'm often getting duplicated information about the help button with NVDA and JAWS with the Iframe editor. It reads something like:

1. Rich Text Editor, [editor name]
2. Editor [editor name] Frame
3. Press alt + 0 for help.
4. Editor [editor name] document
5. Press alt + 0 for help.

I'm not sure if it's valid from the accessibility point of view, but maybe it makes sense to read it only once?

I've also run into a bug where NVDA is reading labels literally twice - but I'm not sure if it's caused by a screen reader or the editor. Did you notice that issue during your tests?

I'm not fully confident that we should read so much information. I understand that technically it may make the editor less accessible by reading only some part of the editor (e.g. only frame) but it's a lot of information for a use case when the focus goes straight to editable. Anyway, I don't feel enough confident about how often there is a case when labeling all parts of the editor is important (and what % of our users it covers), so I'm leaving that up to @Comandeer.

Comandeer commented on tests/core/creators/manual/themeduilabels.md:
```diff
@@ -0,0 +1,23 @@
+Sample announcements for the `iframe`-based editor:
+
+* VoiceOver: "edit text, &lt;content and selection info&gt;. Editor, wysiwygarea, group"
+* NVDA: "Rich Text Editor, wysiwygarea, Editor, wysiwygarea frame, Editor, wysiwygarea document editable"
```
Hmm, did you test NVDA with Chrome? I can see that in Chrome it's read the way you described but in Firefox – the way I described.

It seems that it would be better to annotate browsers used for each screen reader.

Comandeer commented:
> I've also run into a bug where NVDA is reading labels literally twice - but I'm not sure if it's caused by a screen reader or the editor. Did you notice that issue during your tests?

The issue was not present in Firefox and in Chrome it's present only after refocusing the editor therefore I didn't catch it :/ It seems to be a screen reader error as the HTML code is correct and it works correctly in Firefox.

> I'm not sure if it's valid from the accessibility point of view, but maybe it makes sense to read it only once?

That's another thing that does not seem to be fixable… This info is added only once, as the `[aria-describedby]` attribute on the `iframe`, however, Firefox exposes the same description on both the `iframe` itself and the document inside the frame.

Comandeer commented:
I've updated tests and API docs for `config.applicationTitle`. Unfortunately, other raised issues don't seem to be fixable on our side.

jacekbogdanski commented on core/editor.js:
```diff
@@ -1867,15 +1883,51 @@ CKEDITOR.ELEMENT_MODE_INLINE = 3;
  */
 
+/**
+ * Customizes the {@link CKEDITOR.editor#applicationTtle human-readable title} of the application for this
+ * editor. This title is used as a label for the whole website's region containing editor with its toolbars and other
+ * UI parts. This title impacts various
+ * [accessibility aspects](#!/guide/dev_a11y-section-announcing-the-editor-on-the-page),
+ * e.g. it is commonly used by screen readers for distinguishing editor instances and for navigation.
+ * Accepted values are a string or `false`.
```
```suggestion
 * Customizes the {@link CKEDITOR.editor#applicationTtle human-readable title} of the application for this
 * editor. This title is used as a label for the whole website's region containing the editor with its toolbars and other
 * UI parts. Application title impacts various
 * [accessibility aspects](#!/guide/dev_a11y-section-announcing-the-editor-on-the-page),
 * e.g. it is commonly used by screen readers for distinguishing editor instances and for navigation.
 * Accepted values are a string or `false`.
```

jacekbogdanski reviewed (changes requested):
Looks very well, and the explanation given in https://github.com/ckeditor/ckeditor4/issues/2445#issuecomment-1115068142 should make it clear why we choose that approach.

Could you add a manual test with:
* disabled or changed `config.applicationTitle` property
* disabled or changed `config.title` property?

Comandeer commented:
I've added manual tests for labels, six in total:

*   for disabled labels:
    *   `themeduidisabledapplicationlabels`,
    *   `themeduidisablededitorlabels`,
    *   `inlinedisabledapplicationlabels`,
    *   `inlinedisablededitorlabels`,
*   for custom labels:
    *   `themeduicustomlabels`,
    *   `inlinecustomlabels`.

I've also found a bug when an inline editable got the `[aria-label]` with the `false` string as a value instead of not getting this attribute at all. Fixed it and added unit tests for disabled labels as well.

Comandeer commented:
I've pushed one more fix.

The `iframe` has the `title` element with the same label as the one in `iframe[title]` (it seems that this fallback is needed for IE). However, if there is no `[title]` attribute on the `iframe` the `title`'s content is set to `null` string. I've changed it to `&nbsp;` (non-breaking space). I didn't remove the `title` completely as it would cause the screen reader to read the `iframe`'s source URL, making the whole experience far less accessible.

jacekbogdanski reviewed (approved):
Good job @Comandeer, these changes give much better control over editor accessibility compliance for integrators.
